Release management proposal (#600)
submariner-io/submariner: docs/release-management.md

   the whole release process.
   
   Yaml files in this repository can be used to control the definition of a release, 
   
   ```
      releases/0.5-<name>/0.5.0-<name>.yaml
      releases/0.5-<name>/0.5.1-bugfix for....yaml
      ...
   ```

## A release yaml

   A release yaml can have the following format:
   
   ```yaml
    ---
    version: 0.5.0
    name: the mighty 0.5 release!
    release-notes: 
      - bugfixes: >
        * we fixed this bug
      - features: >
        * this feature was added
        * and this other one
      - upgrades: >  # optional
        Upgrade notes important to anybody upgrading from a previous version
      - deprecations: > # optional
        Features or settings that have been deprecated and need to be updated
        by the administrator.

    components:
      - submariner: <commit-id> 
      - lighthouse: <commit-id>
      - submariner-operator: <commit-id>
      - submariner-charts: <commit-id>
      - shipyard: <commit-id>

   ```

## verification of the release yaml

The verification step would:
* Verify the format of the yaml file
* Verify that passed CI runs exist for the commit-ids proposed for the components

> * Verify that submariner-operator dependencies to projects in go.mod are pointing

Author: One alternative is  that all are versions are aligned, so for a 0.5.0 release of the operator we expect all the other to be the same, ... hence we  don't need a versions.go file.

That's more aligned with the yaml format we are defining.

# Proposal for release management in the submariner-io project

## Introduction

Releasing submariner consists of several manual [steps](https://submariner-io.github.io/contributing/release-process/)
that need to be carefully executed in order across multiple projects.

This process can be automated as described in the following sections.

## What's a release

   A submariner release is the sum of all components, tagged on a specific set of versions,
   that work and are tested together.
   
   The components include:
   
   * subctl
   * submariner-operator
   * submariner core (submariner engine, route agent, globalnet ipam)

>    * lighthouse (lighthouse agent, lighthouse coredns)

Reviewer: Do you want to have Shipyard on the list? You mention it later so it can be listed here with a note that it's the base env for building the other projects.

   
  
## A release repository

   A release repository (submariner-io/releases) can be used to track and automate
   the whole release process.
   
   Yaml files in this repository can be used to control the definition of a release, 
   
   ```
      releases/0.5-<name>/0.5.0-<name>.yaml

>       releases/0.5-<name>/0.5.1-bugfix for....yaml

Reviewer: BTW space in filenames is evil :P

      ...
   ```

## A release yaml

   A release yaml can have the following format:
   
   ```yaml
    ---
    version: 0.5.0

>     name: the mighty 0.5 release!

Reviewer: This name becomes the release name in GH? :)

Author: Correct, that'd be the idea.

    release-notes: 
      - bugfixes: >
        * we fixed this bug
      - features: >
        * this feature was added
        * and this other one
      - upgrades: >  # optional
        Upgrade notes important to anybody upgrading from a previous version
      - deprecations: > # optional
        Features or settings that have been deprecated and need to be updated
        by the administrator.

    components:
      - submariner: <commit-id> 
      - lighthouse: <commit-id>
      - submariner-operator: <commit-id>
      - submariner-charts: <commit-id>
      - shipyard: <commit-id>
      - admiral: <commit-id>

   ```

## verification of the release yaml

The verification step would:

> * Verify the format of the yaml file

Reviewer: Not sure if you intend to do it later, but it would be good to clarify what is expected (ie are you just expecting a valid yaml or are you expecting certain fields to be set to certain values?

Author: I mean valid yaml, with the fields we need.

> * Verify that passed CI runs exist for the commit-ids proposed for the components

Reviewer: This is also vague and needs more clarification, e.g are you expecting all jobs that ran to have passed?
Now that we have a lot of CI, and some flaky jobs, perhaps we don't want all CI but just certain jobs (or even certain jobs per repo).

Author: Good point, probably we need a global config (another yaml?) that specifies what jobs are we looking for exactly, like github does.

* Verify that submariner-operator dependencies to projects in go.mod are pointing
  to the proposed submariner, lighthouse, and shipyard releases, because this affects
  the behaviour of `subctl verify` to be up-to-date with the e2e tests of those repositories.

>   Otherwise propose a PR for anything missing and stop.

Reviewer: IDK if this should also propose a PR, since the release would have to be updated anyway after the changes were made.
For starters I'd just fail the release and explain what's the problem.

Author: Makes sense, we can improve later.

* Verify that projects have Dockerfile.dapper pinned to the specific shipyard release,

>   otherwise send a PR to the projects which didn't pin the image yet and stop.

Reviewer: I wouldn't do it here, I'd have another tool (not necessarily a bot) do this.

To me seems that just complaining is fine

Author: Hmmm, it's a lot of manual job. I remember sweating to go through all those changes in repos, it would be nice to automate. If you want we can document something like this as ideas for later work items. (same for L76) ...

May be it could be a completely separate process, or even just some bash/go scripts (like what I made for unpinning).

* Verify that submariner/lighthouse are using the specific admiral commit ID, otherwise send

>   a PR to submariner/lighthouse importing the specific commit ID, and stop.

Reviewer: Again, IDK if I'd send a PR, for starters lets have it just fail and complain.

Author: I think automating those parts are also important, they are probably the biggest gross of the work last time. Probably we should have a big plan, and define iterative work items.

> * Run E2E for the combination of commit-ids with helm & operator (with lighthouse enabled)

Reviewer: This is conditional on the previous verifications passing?

Author: I was assuming everything was conditional on the previous.

## merge of the release yaml

* A script will run in Github actions, executing step by step the tagging and verification
of the multiple components.

It will retry builds if necessary.

> * Shipyard will be tagged for reference

Reviewer: Don't see a point to do this, since we already expect all repos to be pinned to something which would require the SHA (or tag) to exist before the release.

Also, don't you mean we would tag all the repositories (including Shipyard if we want)?

> * The release notes will be added to the website, proposed as a PR.

Reviewer: Would it also create a release of subctl?

> * Shipyard Dockerfile.dapper will be unpinned from all repositories in master via PR.

Reviewer: Coolness :)

## Handling of versions.go in subctl/operator

Subctl and the operator have a set of version references about the other project they manage,
this is handled by updating a versions.go in the repository. For this management strategy
this will need to change in the following way:

### "latest" reference

versions.go will reference latest by default in the source code, but will also work 
as a placeholder for anybody compiling subctl with:

* `-X github.com/submariner-io/submariner-operator/pkg/versions.LighthouseVersion=latest`
* `-X github.com/submariner-io/submariner-operator/pkg/versions.SubmarinerVersion=latest`
* `-X github.com/submariner-io/submariner-operator/pkg/versions.OperatorVersion=latest`

### references pinned at compilation

#### For a master build

For the [devel bleeding edge](https://github.com/submariner-io/submariner-operator/releases/tag/devel)
release channel, every component will be pinned at compilation time to the last successfully
published image of each component (race conditions apart we assume this is what we tested

> in E2E). 

Reviewer: Is this version for reference only, or does it affect the image the subctl actually deploys?

Reviewer: Yeah that would be the simplest

      - submariner-charts: <commit-id>
      - shipyard: <commit-id>
      - admiral: <commit-id>

   ```

## verification of the release yaml

The verification step would:
* Verify the format of the yaml file, it should be valid yaml and have the required
  fields to execute the release.

* Verify that passed CI runs exist for the commit-ids proposed for the components,
  we may have another yaml file in the releases repo defining the jobs that we consider
  stable, as sometimes we have less stable jobs because of new features.

* Verify that submariner-operator dependencies to projects in go.mod are pointing
  to the proposed submariner, lighthouse, and shipyard releases, because this affects
  the behaviour of `subctl verify` to be up-to-date with the e2e tests of those repositories.

* Verify that projects have Dockerfile.dapper pinned to the specific shipyard release.

* Verify that submariner/lighthouse are using the specific admiral commit ID, otherwise stop.

## merge of the release yaml

* A script will run in Github actions, executing step by step the tagging and verification
of the multiple components.

It will retry builds if necessary.

> * Shipyard & Admiral will be tagged for reference

Reviewer: I don't think this would make sense, as we're stating that we're expecting them to be "ready" already before the release.

i.e.
```
Verify that submariner/lighthouse are using the specific admiral commit ID, otherwise stop.
```
If we tag admiral, dependabot will sync it automatically otherwise someone will have to manually sync.

```
Verify that projects have Dockerfile.dapper pinned to the specific shipyard release.
```
Unless you want them pinned to an arbitrary image version (e.g. `v0.5.0-8-gdf0d54`) beforehand, it would be simpler to have them point to a more coherent tag.

Perhaps release can be a 2 phase approach, one would be a "pre-release" stage upon which these are tagged with -preX (X would be some number that we determine either manually or auto incremented), and then automation could pin everything to that version.

Then the actual "release" where everything else that's in this doc make sense.

We could add this in the future of course, doesnt have to happen now, but this seems like a more coherent way of doing it unless you want to have still some manual steps involved in the process (besides the writing of the release yaml of course).

Author: I have tried dependabot in the past and didn't work, may be because I didn't use a "final" release and was 0.5.0-pre0, may be we should start releasing 0.x.0 0.x.1 0.x.2 for admiral by the end of a sprint.

this is handled by updating a versions.go in the repository. For this management strategy
this will need to change in the following way:

### "devel" reference

versions.go will reference devel by default in the source code, but will also work
as a placeholder for anybody compiling subctl with:

* `-X github.com/submariner-io/submariner-operator/pkg/versions.LighthouseVersion=devel`
* `-X github.com/submariner-io/submariner-operator/pkg/versions.SubmarinerVersion=devel`
* `-X github.com/submariner-io/submariner-operator/pkg/versions.OperatorVersion=devel`

### references pinned at compilation

#### For a master build

For the [devel bleeding edge](https://github.com/submariner-io/submariner-operator/releases/tag/devel)
release channel, every component will be pinned at compilation time to the last successfully
published image of each component (race conditions apart we assume this is what we tested
in E2E). 

* `-X github.com/submariner-io/submariner-operator/pkg/versions.LighthouseVersion=${LH_IMAGE_ID}`
* `-X github.com/submariner-io/submariner-operator/pkg/versions.SubmarinerVersion=${SUBM_IMAGE_ID}`

> * `-X github.com/submariner-io/submariner-operator/pkg/versions.OperatorVersion=${OPERATOR_IMAGE_ID}`

Author: We won't have control over the specific versions anymore, so I say let's use just "devel" here.

# Proposal for release management in the submariner-io project

## Introduction

> Releasing submariner contains numerous manual [steps](https://submariner-io.github.io/contributing/release-process/)
> that need to be executed carefully in order accross multiple projects,
> waiting for CI executions to properly happen.

Reviewer 2: ```suggestion
Releasing submariner consists of several manual [steps](https://submariner-io.github.io/contributing/release-process/)
that need to be carefully executed in order across multiple projects.
```

> All this process could be automated as described in the following sections.

Reviewer 2: ```suggestion
This process can be automated as described in the following sections.
```

## What's a release

   A submariner release is the sum of all components, tagged on a specific set of versions,

>    that works, and it's tested together.

Reviewer 2: ```suggestion
   that work and are tested together.
```

   
   The components include:
   
   * subctl
   * submariner-operator
   * submariner core (submariner engine, route agent, globalnet ipam)
   * lighthouse (lighthouse agent, lighthouse coredns)
   
  
## A release repository

>    A release repository (submariner-io/releases) could be used to track and automate
>    all the release process.

Reviewer 2: ```suggestion
   A release repository (submariner-io/releases) can be used to track and automate
   the whole release process.
```

>    Yaml files could be used inside this repository to control the definition of a release, 

Reviewer 2: ```suggestion
   Yaml files in this repository ccan be used to control the definition of a release, 
```

   
   ```
      releases/0.5-<name>/0.5.0-<name>.yaml
      releases/0.5-<name>/0.5.1-bugfix for....yaml
      ...
   ```

## A release yaml

>    A release yaml could have the following format:

Reviewer 2: ```suggestion
   A release yaml can have the following format:
```

   
   ```yaml
    ---
    version: 0.5.0
    name: the mighty 0.5 release!
    release-notes: 
      - bugfixes: >
        * we fixed this bug
      - features: >
        * this feature was added
        * and this other one
      - upgrades: >  # optional
        Upgrade notes important to anybody upgrading from a previous version
      - deprecations: > # optional
        Features or settings that have been deprecated and need to be updated by the administrator.

    components:
      - submariner: <commit-id> 
      - lighthouse: <commit-id>
      - submariner-operator: <commit-id>
      - submariner-charts: <commit-id>
      - shipyard: <commit-id>

   ```

## verification of the release yaml

The verification step would:
* Verify the format of the yaml file
* Verify that passed CI runs exist for the commit-ids proposed for the components
* Run E2E for the combination of commit-ids with helm & operator (with lighthouse enabled)

## merge of the release yaml

> * A script will run in Github actions, executing step by step the tagging, and watching
> of the multiple components.

Reviewer 2: ```suggestion
* A script will run in Github actions, executing step by step the tagging and verification
of the multiple components.
```

It will retry builds if necessary.

> * For the operator repository, a new commit will be automatically pushed with the versions.go

Reviewer: But then you're not using the commit-id you originally had, you'll have another commit id since it's a new patch..

Author: Yeah, that is fine, we would know about it internally, and release on the new.

Author: The problem would be if something merges in the meanwhile... but hopefully we can control ourselves not to merge stuff.. and we can add a verification that our versions patch is being applied on top of the requested commit id

Author: But it's true, the release.yaml would not match with that...commit id, with it's itching, not sure if we can fix that somehow. May be inject those versions as env variables and have some defaults? The problem I see with that is that then.. if somebody tries to develop with subctl they need to use the right env vars... it would be tricky.

We could have subctl may be download the manifest?, or just assume that the versions are all equal to his own version by default?

Reviewer: I think a more sensible approach would be to bump the versions POST the release, so that the master always had the next version (as is done in java maven builds, for example).

This way the hashes dont change and overall the flow is a bit simpler I guess..

Author: Hmm, that'd make sense, I like it, I will update the document.

Reviewer: Cool, in fact the automated release process can bump these version and propose that as a PR to the necessary places :)

Author: I'll update and let's review
